streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala

```diff
@@ -35,6 +35,73 @@ class ZStream[-R, +E, -M, +B, +A](
       } yield Control(pull, Command.noop)
     }
 
+  /**
+   * Returns a stream made of the concatenation in strict order of all the streams
+   * produced by passing each element of this stream to `f0`
+   *
+   * @tparam R1 the type of requirement
+   * @tparam E1 the type of errors
+   * @tparam M1 the type of commands
+   * @tparam B1 the type of marker
+   * @tparam C the type of values
+   * @param f0 a function that yields a new stream for each value produced by this stream
+   * @return a stream made of the concatenation in strict order of all the streams
+   */
+  final def flatMap[R1 <: R, E1 >: E, M1 <: M, B1 >: B, C](
+    f0: A => ZStream[R1, E1, M1, B1, C]
+  ): ZStream[R1, E1, M1, B1, C] = {
+    def go(
+      as: ZIO[R1, Either[E1, B1], A],
+      finalizer: Ref[Exit[_, _] => URIO[R1, _]],
+      currPull: Ref[Option[ZIO[R1, Either[E1, B1], C]]],
+      currCmd: Ref[Option[M1 => ZIO[R1, E1, Any]]]
+    ): ZIO[R1, Either[E1, B1], C] = {
+      val pullOuter: ZIO[R1, Either[E1, B1], Unit] = ZIO.uninterruptibleMask { restore =>
+        restore(as).flatMap { a =>
+          (for {
+            reservation <- f0(a).process.reserve
+            control     <- restore(reservation.acquire)
+            _           <- finalizer.set(reservation.release)
+            _           <- currPull.set(Some(control.pull))
+            _           <- currCmd.set(Some(control.command))
+          } yield ())
+        }
+      }
+
+      currPull.get.flatMap {
+        case None =>
+          pullOuter.foldCauseM(
+            _.failureOrCause match {
+              case Left(Left(e))  => Pull.failNow(e)
```

Comment:
It'd also be nice in the implementation of the above to not use `failureOrCause`, as that potentially "loses" other content in the `Cause`.

Reply:
We will still be losing information with `Cause.sequenceCauseEither`, as soon as you have an `B` you must drop everything else.

Reply:
Hmm, in `sequenceOption` it was dropping the information that there had been a `None` along the way, but I'm not sure I can drop the `B` here.